src/renderer/dx/DxRenderer.cpp

```diff
@@ -154,18 +152,30 @@ DxEngine::~DxEngine()
         D3D_FEATURE_LEVEL_9_1,
     };
 
-    RETURN_IF_FAILED(D3D11CreateDevice(_dxgiAdapter1.Get(),
-                                       D3D_DRIVER_TYPE_UNKNOWN,
-                                       NULL,
-                                       DeviceFlags,
-                                       FeatureLevels,
-                                       ARRAYSIZE(FeatureLevels),
-                                       D3D11_SDK_VERSION,
-                                       &_d3dDevice,
-                                       NULL,
-                                       &_d3dDeviceContext));
-
-    RETURN_IF_FAILED(_dxgiAdapter1->EnumOutputs(0, &_dxgiOutput));
+    auto hardware_result = D3D11CreateDevice(NULL,
+                                             D3D_DRIVER_TYPE_HARDWARE,
+                                             NULL,
+                                             DeviceFlags,
+                                             FeatureLevels,
+                                             ARRAYSIZE(FeatureLevels),
+                                             D3D11_SDK_VERSION,
+                                             &_d3dDevice,
+                                             NULL,
+                                             &_d3dDeviceContext);
+
+    if (FAILED(hardware_result))

```

Comment:
Should we maybe log the hardware failure (`LOG_IF_FAILED`) or perhaps only try software under certain failure codes?

Is there a specific and predictable failure code that means "I don't have hardware because I'm in a VM?" Are there any other well-known failures here that might warrant not attempting again with WARP?

Reply:
I believe it just returns `E_INVALIDARG` which I don't think is sufficient to differentiate.

Reply:
Fair. I'm fine with it then.